### PR TITLE
Improve deploy --cancel output

### LINKED ISSUE #4642: Deployment cancel message is confusing
```
$ oc get pods
NAME             READY     STATUS             RESTARTS   AGE
shell-1-deploy   0/1       DeadlineExceeded   0          6m
shell-2-deploy   0/1       DeadlineExceeded   0          1m
shell-3-c9wov    0/1       Running            4          56s
[vagrant@openshiftdev origin]$ oc deploy shell --cancel
no active deployments to cancel
```

```
{"kind":"DeploymentConfig","apiVersion":"v1","metadata":{"name":"shell","namespace":"default","selfLink":"/oapi/v1/namespaces/default/deploymentconfigs/shell","uid":"c962ffe4-5904-11e5-98c8-080027c5bfa9","resourceVersion":"408","creationTimestamp":"2015-09-12T04:14:39Z","labels":{"run":"shell"}},"spec":{"strategy":{"type":"Rolling","rollingParams":{"updatePeriodSeconds":1,"intervalSeconds":1,"timeoutSeconds":600,"maxUnavailable":"25%","maxSurge":"25%"},"resources":{}},"triggers":[{"type":"ConfigChange"}],"replicas":1,"selector":{"run":"shell"},"template":{"metadata":{"creationTimestamp":null,"labels":{"run":"shell"}},"spec":{"volumes":[{"name":"volume-dg6hh","emptyDir":{}},{"name":"volume-l03xi","emptyDir":{}}],"containers":[{"name":"shell","image":"nginx","resources":{},"volumeMounts":[{"name":"volume-dg6hh","mountPath":"/var/cache/nginx"},{"name":"volume-l03xi","mountPath":"/var/log/nginx"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","dnsPolicy":"ClusterFirst"}}},"status":{"latestVersion":3,"details":{"causes":[{"type":"ConfigChange"}]}}}
I0912 04:21:54.178402   11379 debugging.go:98] GET https://10.1.2.2:8443/api/v1/namespaces/default/replicationcontrollers?labelSelector=openshift.io%2Fdeployment-config.name%3Dshell
I0912 04:21:54.178427   11379 debugging.go:105] Request Headers:
I0912 04:21:54.178433   11379 debugging.go:108]     User-Agent: oc/v1.1.0 (linux/amd64) kubernetes/44c91b1
I0912 04:21:54.183768   11379 debugging.go:123] Response Status: 200 OK in 5 milliseconds
I0912 04:21:54.183789   11379 debugging.go:126] Response Headers:
I0912 04:21:54.183794   11379 debugging.go:129]     Content-Type: application/json
I0912 04:21:54.183799   11379 debugging.go:129]     Date: Sat, 12 Sep 2015 04:21:54 GMT
I0912 04:21:54.183804   11379 debugging.go:129]     Cache-Control: no-store
I0912 04:21:54.183978   11379 request.go:780] Response Body: {"kind":"ReplicationControllerList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/default/replicationcontrollers","resourceVersion":"479"},"items":[{"metadata":{"name":"shell-1","namespace":"default","selfLink":"/api/v1/namespaces/default/replicationcontrollers/shell-1","uid":"c96a5b5d-5904-11e5-98c8-080027c5bfa9","resourceVersion":"352","generation":3,"creationTimestamp":"2015-09-12T04:14:39Z","labels":{"openshift.io/deployment-config.name":"shell","run":"shell"},"annotations":{"kubectl.kubernetes.io/desired-replicas":"1","openshift.io/deployer-pod.name":"shell-1-deploy","openshift.io/deployment-config.latest-version":"1","openshift.io/deployment-config.name":"shell","openshift.io/deployment.cancelled":"true","openshift.io/deployment.phase":"Failed","openshift.io/deployment.status-reason":"The deployment was cancelled by the user","openshift.io/encoded-deployment-config":"{\"kind\":\"DeploymentConfig\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"shell\",\"namespace\":\"default\",\"selfLink\":\"/oapi/v1/namespaces/default/deploymentconfigs/shell\",\"uid\":\"c962ffe4-5904-11e5-98c8-080027c5bfa9\",\"resourceVersion\":\"177\",\"creationTimestamp\":\"2015-09-12T04:14:39Z\",\"labels\":{\"run\":\"shell\"}},\"spec\":{\"strategy\":{\"type\":\"Rolling\",\"rollingParams\":{\"updatePeriodSeconds\":1,\"intervalSeconds\":1,\"timeoutSeconds\":600,\"maxUnavailable\":\"25%\",\"maxSurge\":\"25%\"},\"resources\":{}},\"triggers\":[{\"type\":\"ConfigChange\"}],\"replicas\":1,\"selector\":{\"run\":\"shell\"},\"template\":{\"metadata\":{\"creationTimestamp\":null,\"labels\":{\"run\":\"shell\"}},\"spec\":{\"containers\":[{\"name\":\"shell\",\"image\":\"nginx\",\"resources\":{},\"terminationMessagePath\":\"/dev/termination-log\",\"imagePullPolicy\":\"IfNotPresent\"}],\"restartPolicy\":\"Always\",\"dnsPolicy\":\"ClusterFirst\"}}},\"status\":{\"latestVersion\":1,\"details\":{\"causes\":[{\"type\":\"ConfigChange\"}]}}}"}},"spec":{"replicas":0,"selector":{"deployment":"shell-1","deploymentconfig":"shell","run":"shell"},"template":{"metadata":{"creationTimestamp":null,"labels":{"deployment":"shell-1","deploymentconfig":"shell","run":"shell"},"annotations":{"openshift.io/deployment-config.latest-version":"1","openshift.io/deployment-config.name":"shell","openshift.io/deployment.name":"shell-1"}},"spec":{"containers":[{"name":"shell","image":"nginx","resources":{},"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","dnsPolicy":"ClusterFirst"}}},"status":{"replicas":0,"observedGeneration":3}},{"metadata":{"name":"shell-2","namespace":"default","selfLink":"/api/v1/namespaces/default/replicationcontrollers/shell-2","uid":"82f93138-5905-11e5-98c8-080027c5bfa9","resourceVersion":"401","generation":3,"creationTimestamp":"2015-09-12T04:19:51Z","labels":{"openshift.io/deployment-config.name":"shell","run":"shell"},"annotations":{"kubectl.kubernetes.io/desired-replicas":"1","openshift.io/deployer-pod.name":"shell-2-deploy","openshift.io/deployment-config.latest-version":"2","openshift.io/deployment-config.name":"shell","openshift.io/deployment.cancelled":"true","openshift.io/deployment.phase":"Failed","openshift.io/deployment.status-reason":"The deployment was cancelled by the user","openshift.io/encoded-deployment-config":"{\"kind\":\"DeploymentConfig\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"shell\",\"namespace\":\"default\",\"selfLink\":\"/oapi/v1/namespaces/default/deploymentconfigs/shell\",\"uid\":\"c962ffe4-5904-11e5-98c8-080027c5bfa9\",\"resourceVersion\":\"331\",\"creationTimestamp\":\"2015-09-12T04:14:39Z\",\"labels\":{\"run\":\"shell\"}},\"spec\":{\"strategy\":{\"type\":\"Rolling\",\"rollingParams\":{\"updatePeriodSeconds\":1,\"intervalSeconds\":1,\"timeoutSeconds\":600,\"maxUnavailable\":\"25%\",\"maxSurge\":\"25%\"},\"resources\":{}},\"triggers\":[{\"type\":\"ConfigChange\"}],\"replicas\":1,\"selector\":{\"run\":\"shell\"},\"template\":{\"metadata\":{\"creationTimestamp\":null,\"labels\":{\"run\":\"shell\"}},\"spec\":{\"volumes\":[{\"name\":\"volume-dg6hh\",\"emptyDir\":{}},{\"name\":\"volume-l03xi\",\"emptyDir\":{}}],\"containers\":[{\"name\":\"shell\",\"image\":\"nginx\",\"resources\":{},\"volumeMounts\":[{\"name\":\"volume-dg6hh\",\"mountPath\":\"/var/cache/nginx\"},{\"name\":\"volume-l03xi\",\"mountPath\":\"/var/log/nginx\"}],\"terminationMessagePath\":\"/dev/termination-log\",\"imagePullPolicy\":\"IfNotPresent\"}],\"restartPolicy\":\"Always\",\"dnsPolicy\":\"ClusterFirst\"}}},\"status\":{\"latestVersion\":2,\"details\":{\"causes\":[{\"type\":\"ConfigChange\"}]}}}"}},"spec":{"replicas":0,"selector":{"deployment":"shell-2","deploymentconfig":"shell","run":"shell"},"template":{"metadata":{"creationTimestamp":null,"labels":{"deployment":"shell-2","deploymentconfig":"shell","run":"shell"},"annotations":{"openshift.io/deployment-config.latest-version":"2","openshift.io/deployment-config.name":"shell","openshift.io/deployment.name":"shell-2"}},"spec":{"volumes":[{"name":"volume-dg6hh","emptyDir":{}},{"name":"volume-l03xi","emptyDir":{}}],"containers":[{"name":"shell","image":"nginx","resources":{},"volumeMounts":[{"name":"volume-dg6hh","mountPath":"/var/cache/nginx"},{"name":"volume-l03xi","mountPath":"/var/log/nginx"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","dnsPolicy":"ClusterFirst"}}},"status":{"replicas":0,"observedGeneration":3}},{"metadata":{"name":"shell-3","namespace":"default","selfLink":"/api/v1/namespaces/default/replicationcontrollers/shell-3","uid":"8c86cbdd-5905-11e5-98c8-080027c5bfa9","resourceVersion":"452","generation":2,"creationTimestamp":"2015-09-12T04:20:07Z","labels":{"openshift.io/deployment-config.name":"shell","run":"shell"},"annotations":{"openshift.io/deployer-pod.name":"shell-3-deploy","openshift.io/deployment-config.latest-version":"3","openshift.io/deployment-config.name":"shell","openshift.io/deployment.phase":"Complete","openshift.io/encoded-deployment-config":"{\"kind\":\"DeploymentConfig\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"shell\",\"namespace\":\"default\",\"selfLink\":\"/oapi/v1/namespaces/default/deploymentconfigs/shell\",\"uid\":\"c962ffe4-5904-11e5-98c8-080027c5bfa9\",\"resourceVersion\":\"408\",\"creationTimestamp\":\"2015-09-12T04:14:39Z\",\"labels\":{\"run\":\"shell\"}},\"spec\":{\"strategy\":{\"type\":\"Rolling\",\"rollingParams\":{\"updatePeriodSeconds\":1,\"intervalSeconds\":1,\"timeoutSeconds\":600,\"maxUnavailable\":\"25%\",\"maxSurge\":\"25%\"},\"resources\":{}},\"triggers\":[{\"type\":\"ConfigChange\"}],\"replicas\":1,\"selector\":{\"run\":\"shell\"},\"template\":{\"metadata\":{\"creationTimestamp\":null,\"labels\":{\"run\":\"shell\"}},\"spec\":{\"volumes\":[{\"name\":\"volume-dg6hh\",\"emptyDir\":{}},{\"name\":\"volume-l03xi\",\"emptyDir\":{}}],\"containers\":[{\"name\":\"shell\",\"image\":\"nginx\",\"resources\":{},\"volumeMounts\":[{\"name\":\"volume-dg6hh\",\"mountPath\":\"/var/cache/nginx\"},{\"name\":\"volume-l03xi\",\"mountPath\":\"/var/log/nginx\"}],\"terminationMessagePath\":\"/dev/termination-log\",\"imagePullPolicy\":\"IfNotPresent\"}],\"restartPolicy\":\"Always\",\"dnsPolicy\":\"ClusterFirst\"}}},\"status\":{\"latestVersion\":3,\"details\":{\"causes\":[{\"type\":\"ConfigChange\"}]}}}"}},"spec":{"replicas":1,"selector":{"deployment":"shell-3","deploymentconfig":"shell","run":"shell"},"template":{"metadata":{"creationTimestamp":null,"labels":{"deployment":"shell-3","deploymentconfig":"shell","run":"shell"},"annotations":{"openshift.io/deployment-config.latest-version":"3","openshift.io/deployment-config.name":"shell","openshift.io/deployment.name":"shell-3"}},"spec":{"volumes":[{"name":"volume-dg6hh","emptyDir":{}},{"name":"volume-l03xi","emptyDir":{}}],"containers":[{"name":"shell","image":"nginx","resources":{},"volumeMounts":[{"name":"volume-dg6hh","mountPath":"/var/cache/nginx"},{"name":"volume-l03xi","mountPath":"/var/log/nginx"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent"}],"restartPolicy":"Always","dnsPolicy":"ClusterFirst"}}},"status":{"replicas":1,"observedGeneration":2}}]}
```

### PR DESCRIPTION
**Pull request by ironcladlou**
Make the deploy --cancel output more accurate and informative.

Fixes #4642

### CODE REVIEW

**Comment by ironcladlou** [comment]
Example output:

```
$ oc deploy deploytester --cancel                                                                                                           
no deployments are in progress (latest deployment #1 failed about an hour ago)

$ oc deploy deploytester2 --cancel
there have been no deployments for test/deploytester2

$ oc deploy deploytester2 --cancel                                                                                                          
no deployments found to cancel
```

File: pkg/cmd/cli/cmd/deploy.go [hunk 1]
```diff
@@ -317,7 +322,18 @@ func (o *DeployOptions) cancel(config *deployapi.DeploymentConfig, out io.Writer
...
 	}
 	if !anyCancelled {
-		fmt.Fprintln(out, "no active deployments to cancel")
+		if latest == nil {
+			// TODO: this could mean that a new deployment is forthcoming but hasn't
+			// yet been created; might not be worth trying to express in this
+			// context.
+			fmt.Fprintf(out, "no deployments are in progress\n")
+		} else {
+			timeAt := strings.ToLower(units.HumanDuration(time.Now().Sub(latest.CreationTimestamp.Time)))
```
Comment (kargakis) [inline pkg/cmd/cli/cmd/deploy.go]: I would prefer moving [formatRelativeTime](https://github.com/openshift/origin/blob/392e9db042c41681f5bbc60e0d4c8febf32ed552/pkg/cmd/cli/describe/helpers.go#L113) probably under `pkg/util/time` or at least export it (we already import its package) and use that instead but you would have to change a couple of other places as well.

File: pkg/cmd/cli/cmd/deploy.go [hunk 2]
```diff
@@ -284,14 +286,17 @@ func (o *DeployOptions) cancel(config *deployapi.DeploymentConfig, out io.Writer
 		return err
 	}
 	if len(deployments.Items) == 0 {
-		fmt.Fprintln(out, "no deployments found to cancel")
+		fmt.Fprintf(out, "there have been no deployments for %s/%s\n", config.Namespace, config.Name)
```
Comment (kargakis) [inline pkg/cmd/cli/cmd/deploy.go]: Shouldn't our messages start with capital letters?

Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Not sure. They have always been lower case here, for whatever reason. @smarterclayton?

**Comment by kargakis** [comment]
LGTM, just nits

File: pkg/cmd/cli/cmd/deploy.go (hunk 1 above)
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Spoke with @kargakis and we decided that there needs to be a separate PR to extract various things here into `pkg/util`.

**Comment by ironcladlou** [comment]
@smarterclayton should also take a look since he filed the original issue.

File: pkg/cmd/cli/cmd/deploy.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: We use lowercase words in human readable messages.

On Sep 15, 2015, at 8:58 AM, Dan Mace <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/deploy.go
<https://github.com/openshift/origin/pull/4669#discussion_r39505925>:

> @@ -284,14 +286,17 @@ func (o *DeployOptions) cancel(config *deployapi.DeploymentConfig, out io.Writer
>  		return err
>  	}
>  	if len(deployments.Items) == 0 {
> -		fmt.Fprintln(out, "no deployments found to cancel")
> +		fmt.Fprintf(out, "there have been no deployments for %s/%s\n", config.Namespace, config.Name)

Not sure. They have always been lower case here, for whatever reason.
@smarterclayton <https://github.com/smarterclayton>?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4669/files#r39505925>.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: Oh, it's because this is used with an error: prefix.  Capitalizing after the colon looks wrong

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: No wait, I'm wrong twice.  Should be capitalized if it starts the line on stdout

File: pkg/cmd/cli/cmd/deploy.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: Not into util, into util/*/ :)

File: pkg/cmd/cli/cmd/deploy.go [hunk 3]
```diff
@@ -317,7 +322,18 @@ func (o *DeployOptions) cancel(config *deployapi.DeploymentConfig, out io.Writer
 		return fmt.Errorf("couldn't cancel deployment %s", strings.Join(failedCancellations, ", "))
 	}
 	if !anyCancelled {
-		fmt.Fprintln(out, "no active deployments to cancel")
+		if latest == nil {
+			// TODO: this could mean that a new deployment is forthcoming but hasn't
+			// yet been created; might not be worth trying to express in this
+			// context.
+			fmt.Fprintf(out, "no deployments are in progress\n")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: No deployment is in progress for deployment config foo (latest #3)

Or similar

File: pkg/cmd/cli/cmd/deploy.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Fixed casing across the board within the command.

File: pkg/cmd/cli/cmd/deploy.go (hunk 3 above)
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: The difference in this particular case is there is no latest deployment. THere's nothing in progress, but there might be soon- we can't say for sure.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/deploy.go]: Be more explicit.  No deployments currently in progress for deployment
config foo (current #3)

On Tue, Sep 15, 2015 at 1:20 PM, Dan Mace <notifications@github.com> wrote:

> In pkg/cmd/cli/cmd/deploy.go
> <https://github.com/openshift/origin/pull/4669#discussion_r39538283>:
>
> > @@ -317,7 +322,18 @@ func (o *DeployOptions) cancel(config *deployapi.DeploymentConfig, out io.Writer
> >  		return fmt.Errorf("couldn't cancel deployment %s", strings.Join(failedCancellations, ", "))
> >  	}
> >  	if !anyCancelled {
> > -		fmt.Fprintln(out, "no active deployments to cancel")
> > +		if latest == nil {
> > +			// TODO: this could mean that a new deployment is forthcoming but hasn't
> > +			// yet been created; might not be worth trying to express in this
> > +			// context.
> > +			fmt.Fprintf(out, "no deployments are in progress\n")
>
> The difference in this particular case is there is no latest deployment.
> THere's nothing in progress, but there might be soon- we can't say for sure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4669/files#r39538283>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: I think I got it right this time. Now it'll always show the latest deployment's status when there are >0 deployments.

**Comment by smarterclayton** [comment]
Looks ok

**Comment by ironcladlou** [comment]
Feedback addressed. [merge]
